### DISCUSSION TITLE
loss function: u-u_0,  u_0 is  an array.  how to deal with it?

### DISCUSSION THREAD

**Discussion by yogo64** [lululxvi/deepxde#1898] [Q&A]
I am trying to solve the equation of tv denoising, there is one item in the loss function that is u-u_0, this u_0 is a noisy image, an array. I don't know how to deal with this situation (I am wondering if I want to convert u_0 into a function when I def pde?)

`x = np.arange(noise_image.shape[0]) 
y = np.arange(noise_image.shape[1])  
func = RegularGridInterpolator((x, y), noise_image, method='linear')

def pde(x, u):     #x=(x1,x2,t)
    du_dx1 = dde.grad.jacobian(u, x, j=0)
    du_dx2 = dde.grad.jacobian(u, x, j=1)
    du_t = dde.grad.jacobian(u, x, j=2)
    grad_u = tf.concat([du_dx1, du_dx2], axis=-1)
    norm_grad_u = tf.norm(grad_u, axis=-1, keepdims=True)  # 梯度的模
    v1 = du_dx1 / norm_grad_u
    v2 = du_dx2 / norm_grad_u
    dv1_dx1 = dde.grad.jacobian(v1, x, j=0)
    dv2_dx2 = dde.grad.jacobian(v2, x, j=1)
    w = -(u-func)
    return du_t - dv1_dx1-dv2_dx2 + w`

I get a typeerror:
TypeError: Expected float32 passed to parameter 'y' of op 'Sub', got <scipy.interpolate._rgi.RegularGridInterpolator object at 0x00000215AD5F00A0> of type 'RegularGridInterpolator' instead. Error: Expected float32, but got <scipy.interpolate._rgi.RegularGridInterpolator object at 0x00000215AD5F00A0> of type 'RegularGridInterpolator'.

**Comment by lululxvi** [comment]
Check the example of inverse PDE problems.

**Reply by yogo64** [reply]
谢谢你的回复，我已经参考了。能不能再麻烦你看看我的代码是否正确？我运行后的损失比较大，我已经看过FAQ并试了改动一些东西，但损失依旧不降。Thank you for your reply, I have referred to it. Could you please check if my code is correct again? My loss after running is relatively large. I have read the FAQ and tried to change some things, but the loss is still not reduced.

`import deepxde as dde
import numpy as np
import tensorflow as tf
from matplotlib import pyplot as plt
from PIL import Image
import scipy.io
"用生成的20*20的图运行"

geom = dde.geometry.Rectangle([0, 0], [20, 20])
timedomain = dde.geometry.TimeDomain(0, 1)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

data = Image.open("D:/TVdata/output_image.jpg")
data = data.convert('L')
noise_image0 = np.array(data)
noise_image = noise_image0.flatten()[:, None]
listxy = [(x, y) for x in range(20) for y in range(20)]
x_y = np.array(listxy)

import scipy as sp
def ex_func(x):
    spline = sp.interpolate.Rbf(
        x_y[:, 0], x_y[:, 1], noise_image, function="thin_plate", smooth=0, epsilon=0
    )
    return spline(x[:, 0:1], x[:, 1:2])
ic = dde.icbc.IC(geomtime, lambda x: ex_func(x), lambda _, on_initial: on_initial)

def pde(x, u, ex):
    du_dx1 = dde.grad.jacobian(u, x, j=0)
    du_dx2 = dde.grad.jacobian(u, x, j=1)
    du_t = dde.grad.jacobian(u, x, j=2)
    grad_u = tf.concat([du_dx1, du_dx2], axis=-1)
    norm_grad_u = tf.norm(grad_u, axis=-1, keepdims=True)  # 梯度的模
    v1 = du_dx1 / norm_grad_u
    v2 = du_dx2 / norm_grad_u
    dv1_dx1 = dde.grad.jacobian(v1, x, j=0)
    dv2_dx2 = dde.grad.jacobian(v2, x, j=1)
    w = -0.8*(u-ex)
    return du_t - dv1_dx1-dv2_dx2 + w

bc = dde.NeumannBC(geomtime, lambda x: 0, lambda _, on_boundary: on_boundary)#定义Neumann边界条件


data = dde.data.TimePDE(
    geomtime,
    pde,
    [bc, ic],
    num_domain=400,
    num_boundary=100,
    num_initial=300,
    # anchors=x0,
    # train_distribution="uniform",
    auxiliary_var_function=ex_func,
)

layer_size = [3] + [110] * 3 + [1]
activation = "tanh"
initializer = "Glorot uniform"
net = dde.maps.FNN(layer_size, activation, initializer)

model = dde.Model(data, net)

model.compile("adam", lr=0.001, loss_weights=[1, 1, 1])
losshistory, train_state = model.train(iterations=5000)  # 每100个epoch输出一次
dde.saveplot(losshistory, train_state, issave=True, isplot=True)`
